xds/internal/resolver/serviceconfig.go

```diff
@@ -83,12 +86,23 @@ func (r *xdsResolver) pruneActiveClusters() {
 // serviceConfigJSON produces a service config in JSON format representing all
 // the clusters referenced in activeClusters.  This includes clusters with zero
 // references, so they must be pruned first.
-func serviceConfigJSON(activeClusters map[string]*clusterInfo) ([]byte, error) {
+func serviceConfigJSON(activeClusters map[string]*clusterInfo, clusterSpecifierPlugins map[string]clusterspecifier.BalancerConfig) ([]byte, error) {
 	// Generate children (all entries in activeClusters).
 	children := make(map[string]xdsChildConfig)
 	for cluster := range activeClusters {
-		children[cluster] = xdsChildConfig{
-			ChildPolicy: newBalancerConfig(cdsName, cdsBalancerConfig{Cluster: cluster}),
+		// Look into cluster specifier plugins, which hasn't had any prefix attached to it's cluster specifier plugin names,
+		// to determine the LB Config if the cluster is a CSP.
+		cspCfg, ok := clusterSpecifierPlugins[strings.TrimPrefix(cluster, clusterSpecifierPluginPrefix)]
+		if ok {
```

Comment:
Are we handling this case correctly (I think we're not):

1. First xDS update contains a cluster specifier plugin for RLS
2. RPC starts that selects that CSP
3. New xDS update removes that CSP
4. When generating this JSON here, we need to maintain the CSP config data and keep the "cluster" it references around until (2) is done

Reply:
Ah, I found a solution that solves this problem and the comment about name collisions on the branch in serviceConfigJSON. I simply added the csp balancer configuration (if the cluster is a csp) to the cluster info. This reuses all the plumbing around active clusters, and also solves the correctness issue on the branch of the cluster specifier, as you can branch on clusterInfo.cspCfg != nil.